source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -424,13 +422,13 @@ Network::FilterStatus Filter::initializeUpstreamConnection() {
 
   // Because we never return open connections to the pool, this should either return a handle while
   // a connection completes or it invokes onPoolFailure inline. Either way, stop iteration.
-  upstream_handle_ = conn_pool->newConnection(*this);
+  upstream_handle_.reset(new TcpConnectionHandle(conn_pool->newConnection(*this)));
```

Comment:
That is super subtle - thanks for the tip.

in the long run I wonder if we should refactor this to not be reentrant - seems like it will cause a bunch of confusion?

Either way fixed and (eventually after lots of fun playing with mocks) I think I have a valid regression test.